### ISSUE TITLE
Default session on userbase

### ISSUE THREAD

**Issue by X4fyr** [sddm/sddm#346]
Hey,

It would be great if there would be a possibillty to have a default session on user base, like in lxdm with the `~/.dmrc` file.
So that there is a selection default that reads the `~/.dmrc` of the selected user.

Greetings, X4fyr

**Comment by jleclanche** [comment]
So preferred session per user. Will have to think about that one, it's not a bad idea, but it doesnt belong in ~/.dmrc really.

**Comment by glaubitz** [comment]
Please, for the love of sanity, do not use the accounts-service-daemon for that. gdm and lightdm have already switched to using it meaning that restoring a user's default session and language settings never work when logging in to a different machine on a network with shared home directories which is very common at school and universities.

Please store user-relevant settings in ~/.dmrc or at least in a way which are host-independent. kdm and mdm (a fork of the old gdm 2.20) are the only display managers available which currently do it, all else use the very short-sighted accounts-service-daemon.

Adrian

**Comment by davidedmundson** [comment]
.dmrc is deprecated for a reason

any encrypted home directory doesn't work.
home drives connected on login are becoming more common.

we've had enough bug reports on that causing problems that we're not doing that.

**Comment by glaubitz** [comment]
accounts-service isn't the proper solution either as it stores the settings per host which is completely unusable on network setups with a shared home directory.

I understand the reasoning behind deprecating .dmrc. However, the current approach is not satisfactory either

Please keep networked environments with shared home directories in mind when implementing this feature.

Thanks,

Adrian

**Comment by davidedmundson** [comment]
Fixing AccountsService isn't out the question.
Also there's a possibility that AccountsService will be replaced with SSSD, which is designed for /huge/ deployments.
https://fedorahosted.org/sssd/

**Comment by int-ua** [comment]
I have two users on a machine that use different DEs.
And it's really uncomfortable having to remember to switch session now as it wasn't necessary before updating to SDDM.

**Comment by glaubitz** [comment]
Exactly, apparently most display managers are developed with single user machines, non-networked machines in mind these days.

It surprises that it seems that none of the gdm, lightdm and sddm developers have tested their display managers for that purpose. I have therefore deployed kdm in our environment. I made modifications to Debian's kdm package to integrate better with systemd and have now a usable display manager until we deploy Debian Stretch and thus this problem comes up again because kdm has been deprecated.

**Comment by kkofler** [comment]
AccountSettings also does NOT support a systemwide default as I had requested in #326 (which was closed as a duplicate of this bug, IMHO incorrectly because those are 2 separate issues).

**Comment by stefarossi** [comment]
Any update on this?

**Comment by romainht** [comment]
I am also looking forward to this feature :+1:

**Comment by Vladimir-csp** [comment]
IMHO, `$HOME/.dmrc` should at least be a fallback in case accountsservice or sssd are not present.
It could also be read after user's home dir is accessible and cached somewhere in /var/lib/sddm  for later use.

**Comment by IlyaBizyaev** [comment]
The only reason that stops me from switching to SDDM is absence of per user desktop environment choice. To me such an option feels like an absolutely necessary part of any modern display manager. Waiting eagerly.

**Comment by chrisvs** [comment]
I totally agree with the previous post. I have 2 DE's (xfce and KDE) and I am using lightDM at the moment as login manager. SDDM gets more and more integrated with KDE but does not provide default session per user ...

**Comment by MeikoLoesch** [comment]
I implemented a solution on my system – well quick'n'dirty hack might be more accurate.
Perhaps this helps some people with an annoying problem or might at least be a starting point:

```
$ cat /usr/share/xsessions/rundmrc.desktop
[Desktop Entry]
Encoding=UTF-8
Type=XSession
Exec=run.dmrc
Name=RunDMRC
Comment=Run session saved in ~/.dmrc
### end of rundmrc.desktop
$ cat /usr/local/bin/run.dmrc
#!/bin/bash
SESSION=default
SESSIONDIR=/usr/share/xsessions/
if [ -f ${HOME}/.dmrc ]; then
        SESSION=$(grep -e "^Session=" ${HOME}/.dmrc | cut -d= -f2)
fi
if [ -f ${SESSIONDIR}/${SESSION}.desktop ]; then
        SESSIONFILE=${SESSIONDIR}/${SESSION}.desktop
else
        SESSIONFILE=${SESSIONDIR}/default.desktop
fi                                                                                                                                             
EXECFILE=$(grep -e "^Exec=" ${SESSIONFILE} | cut -d= -f2)                                                                                      
exec ${EXECFILE} 
### end of run.dmrc                                                                                                              
```
This solution lacks decent error handling and one mahor pitfall is that setting a symlink from /usr/share/xsessions/default.desktop to this solution would certainly prevent login for any user without a valid .dmrc. But together with RememberLastSession=true in sddm.conf I can now set different default sessions for users of my pc.

I have not testes, if this solution works together with an encrypted home and have not testes other ditributions as opensuse 42.3 but the general idea should be possible to adapt, if there are problems. There has to be a moment, when the home-directory is properly available, i.e. when it is mounted and decrypted, so why not chose my preferred session after this moment? The location and format (.dmrc, a desktop file, .xinit, .xsession, …) do not really concern me, that is a thing for others to decide...

**Comment by oblitum** [comment]
Just hit this...

Badly needed :-/

**Comment by davidedmundson** [comment]
What's the usecase that isn't covered by us just remembering the last session?

**Comment by oblitum** [comment]
I just need default WM per user, and thought that `~/.dmrc` would do the trick but didn't, then got here.

**Comment by davidedmundson** [comment]
Remembering the last sesion per user with a default entry in the session combo list "Previous" might be feasible.

**Comment by oblitum** [comment]
@davidedmundson yeah, actually, remembering it per session was the thing I was hoping it did when I started using multiple accounts.

**Comment by neallred** [comment]
@oblitum I have the exact same use case (but with `kde`)! I know it works around `sddm` rather than working through it, but I got a workable solution by appending this to `/etc/sddm/Xsession`

```
 if [ $(whoami) = USERNAME_OF_GUI_USER ] ; then . /etc/X11/Xsession /usr/bin/startkde ; else . /etc/X11/Xsession i3 ; fi
```

**Comment by oblitum** [comment]
@neallred thanks for the tip!

**Comment by JustKiddingCode** [comment]
any updates on this?

**Comment by mercuriete** [comment]
I have the following use case:
1. I have a normal user for enjoy youtube or play videogames that I always use with **plasma** kde.
2. I have another user called work that is for productivity and I always use **i3m** to have a clean screen.

It would be nice to remember the last desktop manager per user like @davidedmundson suggested.

Thanks you for your amazing work.
:+1:

**Comment by Cistoge** [comment]
I am quite sure that it is possible to evaluate ~/.dmrc even with an encrypted home file system. Because if is not, I had to wonder about why I never experienced or heard about evaluation of ~/.dmrc failing on a NFS mounted /home where root does not have permissions to read ~/.dmrc of other users, as long as they did not set world-wide read permissions. So, I guess that for example with lightdm, which supports ~/.dmrc, reading of that file takes place after user authentication has already been completed and with the UID of that user.

**Comment by Cistoge** [comment]
I am just trying to adapt the solution of MeikoLoesch with slight modifications, not in a separate script in /usr/local/bin, but already in the SessionCommand file, which in my case is /etc/sddm/Xsession. While doing that, I found that SessionCommand is executed twice; why is it not only running once?

**Comment by Cistoge** [comment]
Running SessionCommand twice disappeared somehow; I have no idea, where it came from or why it went away, surely I did not call it recursively.

Now I am looking if and how it is possible to replace a session that originally was x11 by a wayland session - any idea?

**Comment by Cistoge** [comment]
I am trying to use openvt to create a new virtual terminal for a Wayland session (because I want to have only one "Last used session" entry in the session selection list, but want to handle X and Wayland sessions) like this:

 exec sudo -E openvt -c 9 -s -w -- su -c "$*"  "${USER}"

but though a new terminal opens, I get the following errors when trying to start a wayland plasma session though $HOME/.dmrc. Any idea?

 startplasmacompositor: Starting up...
dbus-daemon[3080]: [session uid=1000 pid=3080] Activating service name='org.freedesktop.systemd1' requested by ':1.1' (uid=1000 pid=3108 comm="dbus-update-activation-environment --systemd --all")
dbus-daemon[3080]: [session uid=1000 pid=3080] Activated service 'org.freedesktop.systemd1' failed: Process org.freedesktop.systemd1 exited with status 1
dbus-update-activation-environment: warning: error sending to systemd: org.freedesktop.DBus.Error.Spawn.ChildExited: Process org.freedesktop.systemd1 exited with status 1
Error: could not determine $DISPLAY.
No backend specified through command line argument, trying auto resolution
FATAL ERROR: backend failed to initialize, exiting now
startplasmacompositor: Shutting down...
startplasmacompositor: Done.

**Comment by Cistoge** [comment]
Coming from MeikoLoesch's script, I implemented a solution that, if the user did not select to use what is already saved in ~/.dmrc, saves the selected session in ~/.dmrc and otherwise, loads the session saved in that file. Both steps work for X11 and Wayland session.

Of course, before starting the implementation, I had set RememberLastSession to false in /etc/sddm.conf.

I also created a new desktop file, in my case /usr/share/xsessions/00previous.desktop (Starting with 00 because sddm sorts its session selection menu entries according to filenames):

 [Desktop Entry]
Type=Application
Exec=/bin/true
DesktopNames=SAVED
Name=Last used session
Name[de]=Zuletzt benutzte Sitzung
Name[en_GB]=Last used session
Comment=Run session saved in ~/.dmrc
Comment[de]=Starte in ~/.dmrc gespeicherte Sitzung
Comment[en_GB]=Run session saved in ~/.dmrc
 
Additionally, I added a script /etc/sddm/session_wrapper and set it as SessionCommand both for X11 and Wayland, using compiler definitions -DSESSION_COMMAND="/etc/sddm/session-wrapper" and -DWAYLAND_SESSION_COMMAND="/etc/sddm/session-wrapper".

The script looks like this (I had to rename it to session-wrapper.txt for uploading):

[session-wrapper.txt](https://github.com/sddm/sddm/files/3651360/session-wrapper.txt)

If the user chooses to not use what is saved in ~/.dmrc, the script just removes whatever was at that location before and creates a new ~/.dmrc. If the user chooses to use what is saved in ~/.dmrc, but ~/.dmrc is not a readable file, a new file is created and a default session is set, I have used lighdm-xsession for that. If it is a readable file, but has no valid Session= entry, the file is not overwritten and the default is used for the session to start.

Next, my script sets environment variables according to the session that will start at the end.

As the final step, for X11 sessions my script uses startx with the original X11 SessionCommand file, in my case /etc/sddm/Xsession; for Wayland sessions, it just the sources the original Wayland SessionCommand file, in my case /etc/sddm/wayland-session.

To make it possible to use startx, I changed the sddm source code so that a new virtual terminal is not only opened for Wayland sessions, but also for X11 sessions:

[enable_dmrc.patch.txt](https://github.com/sddm/sddm/files/3651492/enable_dmrc.patch.txt)

My solution works, but so far I have found three disadvantages:

1. The sddm X11 login screen on (probably) tty7 is blocked while Wayland sessions are running. It is also blocked if I replace startx with sleep or other commands that run for some time without user interaction, while these commands are running. But with startx, the sddm login screen becomes accessable again as soon as X is running on the new tty. Probably not everybody sees this is a disadvantage, but in my opinion, it is.

2. Switching (using Ctrl-Alt-F keys) between the new terminal and tty7 results in the effect that after ending the X11 session, no automatic switch back to tty7 takes place. This can be solved by adding "chvt 7" after the startx in session-wrapper.

3. Using a new virtual terminal not only for Wayland, but also for X11 makes the bug I described in issue #1200 more serious.

Regards
  Christoph

**Comment by Cistoge** [comment]
I found now that disadvantage 1 and 2 are caused by the same thing and has nothing to do with my solution: When more than one Xorg is running and you switch the virtual terminals using the function keys, this lets some of the Xorg processes crash. If this happens to the Xorg started by sddm, a new Xorg is started immediately.

Maybe this problem only occurs with the intel graphics driver.

**Comment by akorn** [comment]
Wasn't this closed by mistake? As far as I can tell, the commit that closed it was about a different issue.

**Comment by MathisP75** [comment]
Still nothing to this day? After 7 years?

**Comment by p5n** [comment]
I use this one as workaround for dmrc and sddm:
https://aur.archlinux.org/packages/run-dmrc-xsession
https://aur.archlinux.org/cgit/aur.git/tree/?h=run-dmrc-xsession

**Comment by MathisP75** [comment]
Does this work for both Wayland and X11 sessions?

**Comment by p5n** [comment]
did not test with wayland
it just runs executable from $SESSION searching in `/usr/share/xsessions/` and `~/.dmrc`
Probably it can be fixed for wayland

**Comment by abhijithb92** [comment]
I did a workaround by changing a qml file of the theme I use - arcolinux-chili. https://git.fosscommunity.in/abhijith/dotfiles/-/tree/master/sddm/arcolinux-chilli

**Comment by pvalsecc** [comment]
> Does this work for both Wayland and X11 sessions?

Nope, if the desktop entry is in /usr/share/xsessions, it works only for X11 and if it's in /usr/share/wayland-sessions, it works only for Wayland. You have to pick one kind.

**Comment by kkofler** [comment]
Well, you could install the file to both locations, or use hardlinks or (probably) symlinks (unless there is some "security" check disallowing symlinks there).

**Comment by ArrayBolt3** [comment]
So from reading the backlog, I gather than `~/.dmrc` is not an acceptable solution for many, and neither is accounts-service-daemon (though I'm not sure if the concerns previously had with it are still a problem almost a decade later). But the problem is that if a per-host solution isn't acceptable because of the shared home directory situation, and a home-directory-centric solution isn't acceptable because of the encrypted home directory situation, we're kind of sunk because those were the only two options (unless there's another standardized shared bit of data that is applicable both on a local machine and in a network situation). So what do we do now? Maybe just implement both and choose one to override the other?

The Kicksecure project is likely going to need this functionality to work for the user-sysmaint-split feature we're working on once we port to Wayland and SDDM (right now we're on X11 and LightDM). If there's some form of consensus on how this should work, I'm more than happy to write the code to make it work.

**Comment by ArrayBolt3** [comment]
Actually, after some research, I think the feature I need for Kicksecure can be implemented without this. I did look briefly into what would be needed to make this happen and made some progress, but if I work on it, it'll likely be in free time.

**Comment by groszdaniel** [comment]
I don't see what the problem is with storing it in `~/.dmrc`. If the home directory is encrypted, just have "Default/Last used" in the session chooser by default, and _once the user enters his password and the home directory becomes available_ read the last used desktop environment from it, and launch it (or the system's default one if there's none set in `~/.dmrc`). And if the user does choose a session from the menu, after the home directory becomes available, write it into `~/.dmrc` and launch it.

It would work just like expected, like KDM did, even when the same home directory is shared between multiple computers on a network. The only limitation would be that it wouldn't tell you which DE it would launch when you use default selection; but that's not much of a problem: if you not sure which one you used last time, you can always pick one manually from the menu, and 99% of the time you probably know the one you used last time is the one you always use.

Optionally, if the home directory is not encrypted, behave like KDM did, and show the last used DE as soon as the username is entered, instead of the "Default/Last used" text.

**Comment by imshinyu** [comment]
2025 is almost over and we still don't have this simple feature...

**Comment by abhijithb92** [comment]
It is possible to implement this by changing QML files of the themes we use. I had done that for arcolinux-chilli theme back in 2022. If you want I can dig up what was done. The link in my earlier comment does not work sadly.

**Comment by imshinyu** [comment]
> It is possible to implement this by changing QML files of the themes we use. I had done that for arcolinux-chilli theme back in 2022. If you want I can dig up what was done. The link in my earlier comment does not work sadly.
> […](#)

please if u can find it tell me im using lightdm-kde-greeter for the time being but its meh
